# RNAseq Funtional and Network Analysis

There are a huge number of tools for performing further analysis based on your RNAseq Differentially Expressed Gene results. Here are just a few tools that can help explore your data further and gain novel insights.

#### Converting between common gene IDs
- [bioDBnet - Biological Database Network](https://biodbnet-abcc.ncifcrf.gov/db/db2db.php)
- [g:Convert Gene ID conversion](https://biit.cs.ut.ee/gprofiler/convert)

#### Whole gene set ontology annotation and enrichment
- [gProfiler g:Profiler – functional enrichment analysis](https://biit.cs.ut.ee/gprofiler/)
- [KEGG KEGG: Kyoto Encyclopedia of Genes and Genomes](https://www.genome.jp/kegg/)
  - Note that kegg uses some different geneIDs and can take some conversion to get it working properly!
- [NeVOmics GitHub - bioinfproject/bioinfo: NeVOmics](https://github.com/bioinfproject/bioinfo)
- [WebGestalt](http://www.webgestalt.org/)

#### Interaction Networks
- [StringDB](https://string-db.org/)
- [GOnet GOnet](https://tools.dice-database.org/GOnet/)
- Cytoscape 
  - [Bingo plugin  Cytoscape App Store - BiNGO](http://apps.cytoscape.org/apps/bingo)
  - [EnrichmentMap plugin Cytoscape App Store - EnrichmentMap](http://apps.cytoscape.org/apps/enrichmentmap)

#### Other visualisation tools
- [Morpheus Morpheus](https://software.broadinstitute.org/morpheus/)
- [diVenn DiVenn 2.0](https://divenn.tch.harvard.edu/)
- [Complex venn Interactive Venn Diagrams](http://www.interactivenn.net/)

#### Whole packages
- [IPA Ingenuity Pathway Analysis | QIAGEN Digital Insights](https://digitalinsights.qiagen.com/products-overview/discovery-insights-portfolio/analysis-and-visualization/qiagen-ipa/)
- [Shiny-Seq https://szenitha.shinyapps.io/shiny-seq3/](https://szenitha.shinyapps.io/shiny-seq3/)
- [BeavR BEAVR: A Browser-based tool for the Exploration And Visualization of RNAseq data](https://github.com/developerpiru/BEAVR#loading-your-data-into-beavr)

## Exercises

When doing the analysis we generated comparisons between Control, Duplication, and Deletion of the gene region. Pick one comparison to focus on for these questions. (Control vs Duplication, Control vs Deletion, or Deletion vs Duplication).

For these downstream analysis you can use the “annotated” csv file (containing just GeneID, Fold Change & pAdjusted) which will be the most easy to use (This was generated by just selecting the most useful columns from the larger data files)

1. Open your chosen comparison in excel/Sheets (or similar tool). You can choose from upregulated genes, downregulated genes, or the complete file (up and down regulated)
2. Filter the genes to an appropriate number. This could be by P value or fold change or a combination:
    - gProfiler:  ~ 300 genes
    - goNet:       ~300 genes
    - String:       ~100 genes
    - diVenn:       any really!

3. Explore your data using one (or more) of these 4 tools. Try to generate a figure that supports you writing a statement:
        This shows that [Deletion/Duplication] results in [increased/decreased]...

4. Put a figure in the shared document. It doesn’t have to be attractive, a basic screenshot is fine! We just want to see the amount of variation that we can generate within a short time! 

Shared Document: [RNAseq Downstream Analysis - Shared Results](https://docs.google.com/presentation/d/1ZJhtYOjzVINXjKvA-Kcbf0ib-SrIYVs1yq7_iuw1DYc/edit?usp=sharing)
